pallets/sgx-account-linker/src/lib.rs

```diff
@@ -0,0 +1,386 @@
+	use frame_system::{ensure_signed, pallet_prelude::*};
+	use sp_core::{ed25519, sr25519};
+	use sp_std::prelude::*;
+
+	use weights::WeightInfo;
+	pub const EXPIRING_BLOCK_NUMBER_MAX: u32 = 10 * 60 * 24 * 30; // 30 days for 6s per block
+	pub const MAX_ETH_LINKS: usize = 3;
+	pub const MAX_BTC_LINKS: usize = 3;
+	pub const MAX_SUB_LINKS: usize = 3;
+
+	#[derive(Encode, Decode, Clone, Debug, Copy, Eq, PartialEq, TypeInfo)]
+	pub enum NetworkType {
+		Kusama,
+		Polkadot,
+		KusamaParachain(u32),
+		PolkadotParachain(u32),
+	}
+
+	#[derive(Encode, Decode, Clone, Debug, Eq, PartialEq, TypeInfo)]
+	pub enum MultiSignature {
+		Sr25519Signature([u8; 64]),
+		Ed25519Signature([u8; 64]),
+		EcdsaSignature([u8; 65]),
+	}
+
+	#[derive(Encode, Decode, Clone, Debug, Eq, PartialEq, TypeInfo)]
+	pub struct LinkedSubAccount<AccountId> {
+		network_type: NetworkType,
+		account_id: AccountId,
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Ethereum address successfully linked. \[Lintentry account, Ethereum account\]
+		EthAddressLinked(T::AccountId, Vec<u8>),
+		/// BTC address successfully linked. \[Lintentry account, BTC account\]
+		BtcAddressLinked(T::AccountId, Vec<u8>),
+		/// Substrate based address successfully linked. \[Lintentry account, substrate account\]
+		SubAddressLinked(T::AccountId, LinkedSubAccount<T::AccountId>),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		// Cannot recover the signature
+		EcdsaRecoverFailure,
+		// Link request expired
+		LinkRequestExpired,
+		// Provided address mismatch the address recovered from signature recovery
+		UnexpectedAddress,
+		// Unexpected ethereum message length error
+		UnexpectedEthMsgLength,
+		// Invalid BTC address to link
+		InvalidBTCAddress,
+		// Expiration block number is too far away from now
+		InvalidExpiringBlockNumber,
+		// Can't get layer one block number
+		LayerOneBlockNumberNotAvailable,
+		// Signature is wrong
+		WrongSignature,
+		// Expected AccountId is [u8; 32]
+		UnexpectedAccountId,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::storage]
+	#[pallet::getter(fn eth_addresses)]
+	pub(super) type EthereumLink<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, Vec<EthAddress>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn btc_addresses)]
+	pub(super) type BitcoinLink<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AccountId, Vec<Vec<u8>>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn sub_addresses)]
+	pub(super) type SubLink<T: Config> = StorageMap<
+		_,
+		Blake2_128Concat,
+		T::AccountId,
+		Vec<LinkedSubAccount<T::AccountId>>,
```

Comment:
It's not yet a problem but it will be soon: Substrate has updated the AccountId to not implement Default anymore: https://github.com/paritytech/substrate/commit/55cd07a7e22c26932f7bd16b87ea5a7569e38eb4

ValueQuery enforces the Value to implement Default. 

I see the following solutions (not for this PR, but for later):
- Use encoded Account Id (i.e. Vec<u8>)
- Use OptionQuery

Reply:
We ran into the same problem, and we haven't decided what to do yet.

Reply:
yes, we need deal with it after update substrate dependency.